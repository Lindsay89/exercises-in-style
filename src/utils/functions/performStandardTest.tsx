import React from "react";
import { render } from "@testing-library/react";
import { DefaultProps } from "../types/defaultProps";

const performStandardTest = (
  Component: React.FunctionComponent<any>,
  defaultProps: DefaultProps,
  defaultClassName: string
) => {
  const dataTestId = defaultProps["data-testid"] as string;
  it("should render without explode", () => {
    const { getByTestId } = render(<Component {...defaultProps} />);

    expect(getByTestId(dataTestId).id).toMatchSnapshot();
  });

  it('should accept an "id" prop', () => {
    const { getByTestId } = render(<Component {...defaultProps} id="foo-id" />);

    expect(getByTestId(dataTestId).id).toBe("foo-id");
  });

  it("should have default className", () => {
    const { getByTestId } = render(<Component {...defaultProps} />);

    expect(getByTestId(dataTestId).classList?.contains(defaultClassName)).toBe(
      true
    );
  });

  it("should allow adding custom classes", () => {
    const { getByTestId } = render(
      <Component {...defaultProps} className="bar" />
    );

    expect(getByTestId(dataTestId).classList.contains("bar")).toBe(true);
  });

  it("should allow to define custom style", () => {
    const { getByTestId } = render(
      <Component {...defaultProps} style={{ margin: "30px" }} />
    );

    expect(getByTestId(dataTestId).getAttribute("style")).toContain("margin");
  });
};
export default performStandardTest;
