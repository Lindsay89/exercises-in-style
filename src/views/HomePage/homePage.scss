@import "../../utils/variables.scss";

@import "../../utils/fonts.scss";

.eis-homePage {
  background-color: $background-color;
  color: $text-color;
  h1 {
    font-size: $extraLargeFont;
  }
  p {
    font-weight: 700;
    font-size: $largeFont;
  }
  .languages-container {
    display: flex;
    width: auto;
    align-self: flex-end;
    gap: 1rem;
    .language-button {
      min-width: 3rem;
      justify-content: center;
      text-transform: uppercase;
      &.selected {
        background-color: $background-color;
        color: $text-color;
      }
    }
  }

  .line-container {
    position: absolute;
    padding: 0;
    top: 50%;
    left: 50%;
    position: absolute;
    display: flex;
    flex-direction: column;
    justify-content: center;
    width: 2.75rem;
    animation: hithere 1s ease 3;
    .line {
      width: 1.25rem;
      height: 0.125rem;
      background-color: $secondary-color;
      border-radius: 0.125rem;
      &.top,
      &.bottom {
        width: 0.5rem;
      }
      &.top {
        transform: rotate(50deg) translate(0.3125rem, -0.4375rem);
      }
      &.bottom {
        transform: rotate(-50deg) translate(0.3125rem, 0.4375rem);
      }
    }
  }
  &.with-sidebar {
    .line-container {
      top: 1rem;
      left: 1rem;
      transition: all 2s ease;
      .line {
        &.top,
        &.bottom {
          width: 1.25rem;
          transition: all 2s;
        }
        &.top {
          transform: rotate(0deg) translate(0rem, -0.25rem);
        }
        &.bottom {
          transform: rotate(0deg) translate(0rem, 0.25rem);
        }
      }
    }
  }
}

@keyframes hithere {
  30% {
    transform: scale(1.2);
  }
  40%,
  60% {
    transform: rotate(-20deg) scale(1.2);
  }
  50% {
    transform: rotate(20deg) scale(1.2);
  }
  70% {
    transform: rotate(0deg) scale(1.2);
  }
  100% {
    transform: scale(1);
  }
}
