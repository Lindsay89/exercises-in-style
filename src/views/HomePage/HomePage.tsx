import React, { useEffect, useRef, useState } from "react";
import classNames from "classnames";
import { DefaultProps } from "../../utils/types";
import { useTranslation } from "react-i18next";
import { availableLanguages } from "../../i18n";
import { Button, Scaffold, Sidebar } from "../../components";

// styles
import "./homePage.scss";

// Required Props
interface HomePageRequiredProps {}

// Optional Props
interface HomePageOptionalProps extends DefaultProps {}

// Combined required and optional props to build the full prop interface
export interface HomePageProps
  extends HomePageRequiredProps,
    HomePageOptionalProps {}

// use the optional prop interface to define the default props
const defaultProps: HomePageOptionalProps = {
  "data-testid": "eis-homePage",
};

const HomePage: React.FC<HomePageProps> = ({ className, ...rest }) => {
  const classList = classNames("eis-homePage", className);
  const { i18n, t } = useTranslation();
  const [homePageClasses, setHomePageClasses] = useState(classList);
  const [hideSidebar, setHideSidebar] = useState(true);

  const handleLineClasses = () => {
    setHomePageClasses(homePageClasses + " with-sidebar");
    setHideSidebar(false);

    setTimeout(() => document.getElementById("line-container")?.remove(), 2000);
  };

  return (
    <Scaffold hideScrollbar={hideSidebar} className={homePageClasses}>
      <div {...rest}>
        <div className="languages-container">
          {availableLanguages.map((lang) => (
            <Button
              onClick={() => i18n.changeLanguage(lang)}
              key={lang}
              className={
                i18n.language.includes(lang)
                  ? "language-button selected outlined"
                  : "language-button primary"
              }
            >
              {lang}
            </Button>
          ))}
        </div>
        <h1>{t("welcomeText")}</h1>
        <div>
          <p>{t("welcomeDesc")}</p>
        </div>
      </div>

      <Button
        className={"line-container"}
        onClick={handleLineClasses}
        type="primary"
        id="line-container"
      >
        <div className="line top" />
        <div className="line center" />
        <div className="line bottom" />
      </Button>
    </Scaffold>
  );
};

HomePage.defaultProps = defaultProps as Partial<HomePageOptionalProps>;

export default React.memo(HomePage);
