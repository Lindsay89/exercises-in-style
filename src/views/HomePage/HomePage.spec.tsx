import React from "react";
import HomePage from "./HomePage";
import performStandardTest from "../../utils/functions/performStandardTest";

const defaultProps = {
  "data-testid": "eis-homePage",
};

describe("HomePage Component", () => {
  performStandardTest(HomePage, defaultProps, "eis-homePage");
});
