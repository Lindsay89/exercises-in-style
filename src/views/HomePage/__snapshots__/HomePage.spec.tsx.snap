// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage Component should render without explode 1`] = `
<DocumentFragment>
  <div
    class="eis-homePage"
    data-testid="eis-homePage"
  >
    <h1>
      welcomeText
    </h1>
    <div>
      <p>
        welcomeDesc
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`HomePage Component should render without exploding 1`] = `
<div
  class="eis-homePage"
  data-testid="eis-homePage"
>
  <h1>
    welcomeText
  </h1>
  <div>
    <p>
      welcomeDesc
    </p>
  </div>
</div>
`;
